zookeeper-client/zookeeper-client-c/src/mt_adaptor.c

```diff
@@ -502,13 +502,13 @@ int32_t fetch_and_add(volatile int32_t* operand, int incr)
 }
 
 // make sure the static xid is initialized before any threads started
-__attribute__((constructor)) int32_t get_xid()
+int32_t get_xid()
 {
-    static int32_t xid = -1;
-    if (xid == -1) {
-        xid = time(0);
-    }
-    return fetch_and_add(&xid,1);
+    static int32_t xid = 1;
+
+    // The XID returned should not be negative to avoid collisions
+    // with reserved XIDs, such as AUTH_XID or SET_WATCHES_XID.
+    return fetch_and_add(&xid,1) & ~(1<<31);
```

Comment:
This will return `0` when the overflow happens at the first time, so `xid == INT_MIN`, right?
Which is not consistent with the single threaded case where we restart xid from 1 (similarly to the java client).

Reply:
You are correct. -- The first revision of the patch had behavior (and code) that was symmetric between ST/MT. -- I looked at the code (and ran some tests) and there is nothing special about XID being 0 in the C client.

I think at this point we have two options:

1) reset XID to 0 instead of 1 in the ST version of the code. This would give you symmetric behavior (even tho the code no longer looks symmetric).

2) If there is some opposition against ever having 0 returned, I can modify the MT code such that if you were about to return 0, you do another fetch and add. Something like:

    result = 0;
    while (result == 0) {
        result = fetch_and_add(&xid,1) & ~(1<<31);
    }
    return result;

Please LMK what your preference is. IMHO none of the two solutions proposed has a correctness issue.

Reply:
I don't think there's anything wrong with using `0` as `xid`, but I would stick to consistent behaviour between clients anyway.
So, I vote for option 2., but let's see what other people think.